File: leetcode/medium/deepest-leaves-sum.rs
// https://leetcode.com/problems/deepest-leaves-sum
//
// Complexity: runtime O(n), space O(n) [O(logn) if balanced].

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn deepest_leaves_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(
            node: Option<Rc<RefCell<TreeNode>>>,
            depth: i32,
            max_depth: &mut i32,
            sum: &mut i32,
        ) {
            if let Some(node) = node {
                if depth > *max_depth {
                    *max_depth = depth;
                    *sum = 0;
                }

                let node = node.borrow();
                if depth == *max_depth {
                    *sum += node.val;
                }

                dfs(node.left.clone(), depth + 1, max_depth, sum);
                dfs(node.right.clone(), depth + 1, max_depth, sum);
            }
        }

        let (mut max_depth, mut sum) = (0, 0);
        dfs(root, 0, &mut max_depth, &mut sum);

        sum
    }
}

File: leetcode/medium/deepest-leaves-sum.cpp
// https://leetcode.com/problems/deepest-leaves-sum
//
// Complexity: runtime O(n), space O(n) if unbalanced; O(logn) otherwise.

class Solution {
    void dfs(const TreeNode* const root, const int depth, int& maxDepth,
             int& sum) const
    {
        if (!root)
            return;

        if (depth > maxDepth) {
            maxDepth = depth;
            sum = 0;
        }
        if (depth == maxDepth)
            sum += root->val;

        dfs(root->left, depth + 1, maxDepth, sum);
        dfs(root->right, depth + 1, maxDepth, sum);
    }

public:
    int deepestLeavesSum(const TreeNode* const root) const
    {
        int maxDepth = 0, sum = 0;
        dfs(root, 0, maxDepth, sum);
        return sum;
    }
};
